{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import multiprocessing\n",
    "import parmap\n",
    "import datetime\n",
    "import os\n",
    "from pathonoia import pathonoia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this folder will be created and for every dataset and sample corresponding subfolders\n",
    "OUTPUT_DIR = \"./OUTPUT/\"\n",
    "\n",
    "# The Inputfolder should contain one or more fasta files\n",
    "# These files should be samples of the same dataset\n",
    "# if the dataset was originally host RNAseq, align it first to the host's transcriptome and extract the unmapped reads\n",
    "# the fasta files should contain all (host-)unmapped reads \n",
    "# the unmapped reads can be extracted from the alignment BAM file with SAMtools:\n",
    "\n",
    "# samtools view -hb -f 4 host_align.bam > host_unmapped.bam\n",
    "# samtools view host_unmapped.bam | awk \\'{OFS=\"\\\\t\"; print \">\"$1\"\\\\n\"$10}\\' - > host_unmapped.fa\n",
    "\n",
    "inputfolder = \"./dataset/\"\n",
    "\n",
    "# make sure Kraken 2 is installed on your system along with its index build with k=31 AND l=31 \n",
    "# Instructions for creating this index can be found at the end of this notebook\n",
    "krakenTOOL=\"../tools/kraken2/kraken2\"\n",
    "krakenDB=\"../tools/kraken2/db/kraken2_k31\"\n",    
    "# this database is downloadable from oasis.ims.bio/kraken2_k31.zip or oasis.dzne.de/kraken2_31.zip \n",
    "pathonoia.checkKrakenInstall(krakenTOOL,krakenDB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run this cell, these are definitions / functions, only\n",
    "\n",
    "def processDataset(inputfolder):\n",
    "    print(inputfolder)\n",
    "    ds = inputfolder.rsplit(\"/\", 2)[1]\n",
    "    print(ds)\n",
    "    argumentList = []\n",
    "    for file in os.listdir(inputfolder):\n",
    "        if os.path.isfile(inputfolder+file):\n",
    "            if file.endswith(\".fa\") or file.endswith(\".fasta\"):           \n",
    "                sample = file.replace(\".fasta\",\"\").replace(\".fa\",\"\")\n",
    "                odir = OUTPUT_DIR + ds + \"/\" + sample\n",
    "                s1 = inputfolder + file\n",
    "                tup = (ds, odir, sample, s1)\n",
    "                argumentList.append(tup)\n",
    "            else:\n",
    "                print(file + \" is not a fasta file, PASS\")\n",
    "        else:\n",
    "            print(file + \" is not a file! PASS\")\n",
    "            \n",
    "    print(str(len(argumentList)) + \" samples to process\")\n",
    "    \n",
    "    if not pathonoia.checkKrakenInstall(krakenTOOL, krakenDB):\n",
    "        print(\"Problem with Kraken Installation. ABORT\")\n",
    "        return()\n",
    "\n",
    "    allNucPerTaxList = parmap.starmap(runOneSample, argumentList, pm_processes=1, pm_pbar=True)\n",
    "    \n",
    "    print(\"merge and save results\")\n",
    "    allNucPerTax = dict()\n",
    "    for i in range(len(argumentList)):\n",
    "        allNucPerTax[argumentList[i][2]] = allNucPerTaxList[i]\n",
    "\n",
    "    dsdf = pathonoia.mergeNucTaxDataframes(allNucPerTax)\n",
    "    dsdf.to_csv(OUTPUT_DIR + ds + \"/\" + ds + \"_contamination.csv\")\n",
    "    print(ds + \" done\")\n",
    "    return(dsdf)\n",
    "\n",
    "def runOneSample(ds, outputdir, samplename, sample1):   \n",
    "    print(outputdir)\n",
    "    if(os.path.isdir(outputdir)):\n",
    "        print(outputdir + \" ALREADY EXISTS\")\n",
    "    else:\n",
    "        os.makedirs(outputdir)\n",
    "    outdir_prefix = outputdir + \"/\" + samplename\n",
    "    logfile = open(outdir_prefix + \"_pathonoiaLOG_\"+datetime.datetime.now().strftime(\"%Y%m%d\")+\".log\", \"a\") \n",
    "    \n",
    "    print(\"++++++++++++++\", file = logfile)\n",
    "    print(samplename + \"\\n\", file = logfile)\n",
    "    \n",
    "    print(\"Kraken2\", file = logfile)\n",
    "    if (os.path.isfile(outdir_prefix + \"_krakenalign.txt\")):\n",
    "        print(\"Kraken results already available, skip\", file = logfile)\n",
    "    else:\n",
    "        execute(KRAKENcommand(outdir_prefix, sample1),logfile)\n",
    "\n",
    "    print(\"evaluate\", file = logfile)\n",
    "    nucCutOff = 100\n",
    "    print(\"nucCutOff = \" + str(nucCutOff), file = logfile)\n",
    "    nucPerTax = pathonoia.evalKrakenAlign(outdir_prefix, sample1, nucCutOff, logfile)\n",
    "    print(\"DONE\", file = logfile)\n",
    "    logfile.close()\n",
    "    print(samplename + \" done\")\n",
    "    return(nucPerTax)\n",
    "\n",
    "def execute(cmd, logfile):\n",
    "    print(cmd, file = logfile)\n",
    "    answer = \"cmd didnt run, uncommented\"\n",
    "    answer = os.system(cmd)\n",
    "    print(answer, file = logfile)    \n",
    "\n",
    "def KRAKENcommand(outdir_prefix, sample):\n",
    "    #if you want to change settings here, make sure the spaces in the string stay the way they are\n",
    "    krakenPARAMETERS = \"--db \" + krakenDB + \" --use-names --threads 8 --memory-mapping \"\n",
    "    outputPARAMETER = \"--output \" + outdir_prefix + \"_krakenalign.txt --report \" + outdir_prefix + \"_krakenreport.txt \"\n",
    "    samplePARAMETER = sample\n",
    "    cmd = krakenTOOL + \" \" + krakenPARAMETERS + outputPARAMETER + samplePARAMETER\n",
    "    return(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./dataset/\n",
      "dataset\n",
      ".ipynb_checkpoints is not a file! PASS\n",
      "5 samples to process\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "  0%|          | 0/5 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./OUTPUT/dataset/buccal_small\n",
      "buccal_small done\n",
      "./OUTPUT/dataset/cityparks_small\n",
      "cityparks_small done\n",
      "./OUTPUT/dataset/hous1_small\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      " 40%|████      | 2/5 [02:04<03:06, 62.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hous1_small done\n",
      "./OUTPUT/dataset/hous2_small\n",
      "hous2_small done\n",
      "./OUTPUT/dataset/soil_small\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      " 80%|████████  | 4/5 [04:22<01:06, 66.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soil_small done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6it [06:15, 62.57s/it]                       "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "merge and save results\n",
      "dataset done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# This will take time depending on dataset size\n",
    "# The terminal in which the notebook is running may give additional output\n",
    "\n",
    "ds_contamination = processDataset(inputfolder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>phylo_level</th>\n",
       "      <th>parent</th>\n",
       "      <th>buccal_small</th>\n",
       "      <th>cityparks_small</th>\n",
       "      <th>hous1_small</th>\n",
       "      <th>hous2_small</th>\n",
       "      <th>soil_small</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1189310</th>\n",
       "      <td>Myxococcus macrosporus DSM 14697</td>\n",
       "      <td>S1</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1636235.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1632272.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Myxococcus macrosporus</td>\n",
       "      <td>S</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1632307.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1629304.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208963</th>\n",
       "      <td>Pseudomonas aeruginosa UCBPP-PA14</td>\n",
       "      <td>S1</td>\n",
       "      <td>287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1534983.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208964</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>S1</td>\n",
       "      <td>287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1527298.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>S</td>\n",
       "      <td>136841</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1526802.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509170</th>\n",
       "      <td>Acinetobacter baumannii SDF</td>\n",
       "      <td>S1</td>\n",
       "      <td>470</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1007666.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1096997</th>\n",
       "      <td>Acinetobacter baumannii BJAB0868</td>\n",
       "      <td>S1</td>\n",
       "      <td>470</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1007417.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>S</td>\n",
       "      <td>909768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1007309.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1785128</th>\n",
       "      <td>Acinetobacter lactucae</td>\n",
       "      <td>S</td>\n",
       "      <td>909768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>888271.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374930</th>\n",
       "      <td>Haemophilus influenzae PittEE</td>\n",
       "      <td>S1</td>\n",
       "      <td>727</td>\n",
       "      <td>840191.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262727</th>\n",
       "      <td>Haemophilus influenzae R2846</td>\n",
       "      <td>S1</td>\n",
       "      <td>727</td>\n",
       "      <td>840159.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>Haemophilus influenzae</td>\n",
       "      <td>S</td>\n",
       "      <td>724</td>\n",
       "      <td>840048.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246197</th>\n",
       "      <td>Myxococcus xanthus DK 1622</td>\n",
       "      <td>S1</td>\n",
       "      <td>34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>299212.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>301307.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Myxococcus xanthus</td>\n",
       "      <td>S</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>299064.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>301153.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297742</th>\n",
       "      <td>Myxococcus hansupus</td>\n",
       "      <td>S</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>297899.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>299955.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              species_name phylo_level  parent  buccal_small  \\\n",
       "1189310   Myxococcus macrosporus DSM 14697          S1      35           NaN   \n",
       "35                  Myxococcus macrosporus           S      32           NaN   \n",
       "208963   Pseudomonas aeruginosa UCBPP-PA14          S1     287           NaN   \n",
       "208964         Pseudomonas aeruginosa PAO1          S1     287           NaN   \n",
       "287                 Pseudomonas aeruginosa           S  136841           NaN   \n",
       "509170         Acinetobacter baumannii SDF          S1     470           NaN   \n",
       "1096997   Acinetobacter baumannii BJAB0868          S1     470           NaN   \n",
       "470                Acinetobacter baumannii           S  909768           NaN   \n",
       "1785128             Acinetobacter lactucae           S  909768           NaN   \n",
       "374930       Haemophilus influenzae PittEE          S1     727      840191.0   \n",
       "262727        Haemophilus influenzae R2846          S1     727      840159.0   \n",
       "727                 Haemophilus influenzae           S     724      840048.0   \n",
       "246197          Myxococcus xanthus DK 1622          S1      34           NaN   \n",
       "34                      Myxococcus xanthus           S      32           NaN   \n",
       "1297742                Myxococcus hansupus           S      32           NaN   \n",
       "\n",
       "         cityparks_small  hous1_small  hous2_small  soil_small  \n",
       "1189310        1636235.0          NaN          NaN   1632272.0  \n",
       "35             1632307.0          NaN          NaN   1629304.0  \n",
       "208963               NaN    1534983.0          NaN         NaN  \n",
       "208964               NaN    1527298.0          NaN         NaN  \n",
       "287                  NaN    1526802.0          NaN         NaN  \n",
       "509170               NaN          NaN    1007666.0         NaN  \n",
       "1096997              NaN          NaN    1007417.0         NaN  \n",
       "470                  NaN          NaN    1007309.0         NaN  \n",
       "1785128              NaN          NaN     888271.0         NaN  \n",
       "374930               NaN          NaN          NaN         NaN  \n",
       "262727               NaN          NaN          NaN         NaN  \n",
       "727                  NaN          NaN          NaN         NaN  \n",
       "246197          299212.0          NaN          NaN    301307.0  \n",
       "34              299064.0          NaN          NaN    301153.0  \n",
       "1297742         297899.0          NaN          NaN    299955.0  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_contamination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# continue processing the dataset with downstream_analysis notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# setup Kraken 2 for Pathonoia\n",
    "follow these commands one by one to build a Kraken Index that works with Pathonoia\n",
    "\n",
    "    git clone https://github.com/DerrickWood/kraken2.git\n",
    "    cd kraken2/\n",
    "    ./install_kraken2.sh .\n",
    "    mkdir db\n",
    "    ./kraken2-build --download-taxonomy --db db/bacvir_k31 #?--skip-maps\n",
    "    ./kraken2-build --download-library bacteria --db db/bacvir_k31 --use-ftp\n",
    "    ./kraken2-build --download-library viral --db db/bacvir_k31 --use-ftp\n",
    "\n",
    "    ./kraken2-build --build --db db/bacvir_k31 --threads 8 --kmer-len 31 --minimizer-len 31\n",
    "\n",
    "    ./kraken2-build --clean --db db/bacvir_k31"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
